File: src/main/java/me/lokka30/levelledmobs/customdrops/CustomDropsDefaults.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.customdrops;

import java.util.LinkedList;
import java.util.List;
import me.lokka30.levelledmobs.misc.CachedModalList;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Holds all default values for either all custom drop items
 *
 * @author stumper66
 * @since 2.4.0
 */
public class CustomDropsDefaults {
    boolean noMultiplier;
    boolean noSpawner;
    public boolean override;
    boolean playerCausedOnly;
    boolean onlyDropIfEquipped;
    public int amount;
    public int priority;
    public int minLevel;
    public int maxLevel;
    public boolean equipOffhand;
    public boolean equipOnHelmet;
    int customModelData;
    int maxDropGroup;
    int minPlayerLevel;
    int maxPlayerLevel;
    public SlidingChance chance;
    public boolean useChunkKillMax;
    @Nullable SlidingChance equippedChance;
    @Nullable SlidingChance overallChance;
    String groupId;
    String playerLevelVariable;
    String nbtData;
    List<String> itemFlagsStrings;
    public final List<String> permissions;
    final List<String> overallPermissions;
    CachedModalList<DeathCause> causeOfDeathReqs;
    public String externalType;
    public String externalItemId;
    public Double externalAmount;
    public boolean runOnSpawn;
    public boolean runOnDeath;

    CustomDropsDefaults() {
        // these are the defaults of the defaults
        this.amount = 1;
        this.minLevel = -1;
        this.maxLevel = -1;
        this.minPlayerLevel = -1;
        this.maxPlayerLevel = -1;
        this.customModelData = -1;
        this.priority = 0;
        this.maxDropGroup = 0;
        this.noMultiplier = false;
        this.noSpawner = false;
        this.override = false;
        this.playerCausedOnly = false;
        this.permissions = new LinkedList<>();
        this.overallPermissions = new LinkedList<>();
        this.runOnSpawn = false;
        this.runOnDeath = true;
        this.equipOnHelmet = false;
    }

    void setDefaultsFromDropItem(@NotNull final CustomDropBase dropBase) {
        if (this.chance != null)
            this.chance.setFromInstance(dropBase.chance);
        else
            this.chance = dropBase.chance;
        this.useChunkKillMax = dropBase.useChunkKillMax;
        this.amount = dropBase.getAmount();
        this.minLevel = dropBase.minLevel;
        this.maxLevel = dropBase.maxLevel;
        this.priority = dropBase.priority;
        this.maxDropGroup = dropBase.maxDropGroup;
        this.noSpawner = dropBase.noSpawner;
        this.playerCausedOnly = dropBase.playerCausedOnly;
        if (!"default".equals(dropBase.groupId)){
            this.groupId = dropBase.groupId;
        }
        this.minPlayerLevel = dropBase.minPlayerLevel;
        this.maxPlayerLevel = dropBase.maxPlayerLevel;
        this.playerLevelVariable = dropBase.playerLevelVariable;
        this.permissions.addAll(dropBase.permissions);
        this.causeOfDeathReqs = dropBase.causeOfDeathReqs;

        if (dropBase instanceof final CustomDropItem dropItem) {
            this.customModelData = dropItem.customModelDataId;
            if (this.equippedChance != null)
                this.equippedChance.setFromInstance(dropItem.equippedChance);
            else
                this.equippedChance = dropItem.equippedChance;
            this.noMultiplier = dropItem.noMultiplier;
            this.onlyDropIfEquipped = dropItem.onlyDropIfEquipped;
            this.externalType = dropItem.externalType;
            this.externalItemId = dropItem.externalItemId;
            this.externalAmount = dropItem.externalAmount;
            this.equipOffhand = dropItem.equipOffhand;
            this.nbtData = dropItem.nbtData;
            this.itemFlagsStrings = dropItem.itemFlagsStrings;
            this.equipOnHelmet = dropItem.equipOnHelmet;
        } else if (dropBase instanceof final CustomCommand command) {
            this.runOnSpawn = command.runOnSpawn;
            this.runOnDeath = command.runOnDeath;
        }
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/customdrops/CustomDropsDefaults.kt
package io.github.arcaneplugins.levelledmobs.customdrops

import io.github.arcaneplugins.levelledmobs.misc.CachedModalList

/**
 * Holds all default values for either all custom drop items
 *
 * @author stumper66
 * @since 2.4.0
 */
class CustomDropsDefaults {
    var noMultiplier = false
    var noSpawner = false
    var override = false
    var playerCausedOnly = false
    var onlyDropIfEquipped = false
    var amount = 1
    var priority = 0
    var minLevel = -1
    var maxLevel = -1
    var equipOffhand = false
    var equipOnHelmet = false
    var customModelData = -1
    var maxDropGroup = 0
    var minPlayerLevel = -1
    var maxPlayerLevel = -1
    var chance: SlidingChance? = null
    var useChunkKillMax = false
    var equippedChance: SlidingChance? = null
    var overallChance: SlidingChance? = null
    var groupId: String? = null
    var playerLevelVariable: String? = null
    var nbtData: String? = null
    var itemFlagsStrings: MutableList<String>? = null
    var permissions = mutableListOf<String>()
    var overallPermissions = mutableListOf<String>()
    var causeOfDeathReqs: CachedModalList<DeathCause>? = null
    var externalType: String? = null
    var externalItemId: String? = null
    var externalAmount: Double? = null
    var runOnSpawn = false
    var runOnDeath = true

    fun setDefaultsFromDropItem(
    dropBase: CustomDropBase
    ) {
        if (this.chance != null)
            this.chance!!.setFromInstance(dropBase.chance)
        else
            this.chance = dropBase.chance

        this.useChunkKillMax = dropBase.useChunkKillMax
        this.amount = dropBase.amount
        this.minLevel = dropBase.minLevel
        this.maxLevel = dropBase.maxLevel
        this.priority = dropBase.priority
        this.maxDropGroup = dropBase.maxDropGroup
        this.noSpawner = dropBase.noSpawner
        this.playerCausedOnly = dropBase.playerCausedOnly
        if ("default" != dropBase.groupId) {
            this.groupId = dropBase.groupId
        }
        this.minPlayerLevel = dropBase.minPlayerLevel
        this.maxPlayerLevel = dropBase.maxPlayerLevel
        this.playerLevelVariable = dropBase.playerLevelVariable
        permissions.addAll(dropBase.permissions)
        this.causeOfDeathReqs = dropBase.causeOfDeathReqs

        if (dropBase is CustomDropItem) {
            this.customModelData = dropBase.customModelDataId
            if (this.equippedChance != null)
                this.equippedChance!!.setFromInstance(dropBase.equippedChance)
            else
                this.equippedChance = dropBase.equippedChance
            this.noMultiplier = dropBase.noMultiplier
            this.onlyDropIfEquipped = dropBase.onlyDropIfEquipped
            this.externalType = dropBase.externalType
            this.externalItemId = dropBase.externalItemId
            this.externalAmount = dropBase.externalAmount
            this.equipOffhand = dropBase.equipOffhand
            this.nbtData = dropBase.nbtData
            this.itemFlagsStrings = dropBase.itemFlagsStrings
            this.equipOnHelmet = dropBase.equipOnHelmet
        } else if (dropBase is CustomCommand) {
            this.runOnSpawn = dropBase.runOnSpawn
            this.runOnDeath = dropBase.runOnDeath
        }
    }
}